packages/babel-parser/test/helpers/fixture-error.js

```diff
@@ -0,0 +1,84 @@
+import { inspect } from "util";
+import Difference from "./difference.js";
+import "./polyfill.js";
+
+const { isArray } = Array;
+const { defineProperty, entries, fromEntries } = Object;
+
+const named = (name, object) => defineProperty(object, "name", { value: name });
+const mapEntries = (object, f) => fromEntries(entries(object).map(f));
+
+export default class FixtureError extends Error {
+  constructor(message, difference) {
+    super(message);
+    this.difference = difference;
+  }
+
+  static fromDifference(difference, actual) {
+    return difference === Difference.None
+      ? FixtureError.None
+      : difference.path[0] !== "threw"
+      ? new FixtureError.DifferentAST(difference)
+      : !difference.expected
+      ? new FixtureError.UnexpectedError(difference, actual.threw)
+      : difference.actual
+      ? new FixtureError.DifferentError(difference, actual.threw)
+      : actual.ast && actual.ast.errors
+      ? new FixtureError.UnexpectedRecovery(difference, actual.ast.errors)
+      : new FixtureError.UnexpectedSuccess(difference);
+  }
+}
+
+FixtureError.None = Object.freeze(
+  Object.setPrototypeOf({}, FixtureError.prototype),
+);
+
+Object.assign(
+  FixtureError,
+  mapEntries(
+    {
+      DifferentError: ({ expected }) =>
+        `Expected unrecoverable error:    \n\n${expected}\n\n` +
+        `But instead encountered different unrecoverable error.`,
+
+      DifferentAST: ({ message }) => message,
+
+      UnexpectedError: () => `Encountered unexpected unrecoverable error.`,
+
+      UnexpectedSuccess: ({ expected }) =>
+        `Expected unrecoverable error:\n\n    ${expected}\n\n` +
+        `But parsing succeeded without errors.`,
+
+      UnexpectedRecovery: ({ expected }, errors) =>
+        `Expected unrecoverable error:\n\n    ${expected}\n\n` +
+        `But instead parsing recovered from ${errors.length} errors.`,
+    },
+    ([name, toMessage]) => [
+      name,
+      named(
+        name,
+        class extends FixtureError {
+          constructor(difference, cause) {
```

Comment:
Sure, just as an aside, unrelated to this, it's weird that Node seems to have a bug where if you do `new Error("blah", { cause })`, the cause *won't* show up in the stack trace. However, if you do `x = new Error("blah"); x.cause = ...`, then the cause *will* show up in the stack trace. This is because passing cause to the constructor form makes it so cause will not be enumerable, and thus `inspect` seems to miss it. So as of node 16, the only way to see `cause` is to *not* use the standard constructor form.

Reply:
Done: https://github.com/babel/babel/pull/14201/commits/f6aeb0651250a51ec6b0838a6525a8364e96b82c